core/cache/application.go

```diff
@@ -28,18 +33,35 @@ type Application struct {
 	configHash string
 }
 
-func (m *Application) setDetails(details ApplicationChange) {
-	m.mu.Lock()
-	defer m.mu.Unlock()
-	m.details = details
+func (a *Application) WatchFields(comparitors ...func(*ApplicationDelta) bool) *ApplicationFieldWatcher {
+	return newApplicationFieldWatcher(a.hub, a.topic(applicationFieldChange), comparitors)
+}
+
+// TODO (manadart 2018-03-13) Should we change this up and down the call stack
+// so that pointers are passed instead of copies?
+func (a *Application) setDetails(details ApplicationChange) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+
+	delta := &ApplicationDelta{old: a.details, new: details}
+	a.details = details
 
 	configHash, err := hash(details.Config)
 	if err != nil {
 		logger.Errorf("invariant error - application config should be yaml serializable and hashable, %v", err)
 		configHash = ""
 	}
-	if configHash != m.configHash {
-		m.configHash = configHash
+	if configHash != a.configHash {
+		a.configHash = configHash
 		// TODO: publish config change...
 	}
+
+	a.hub.Publish(a.topic(applicationFieldChange), delta)
```

Comment:
does this mean that applications that use large config values (like, say, CA certs), end up with 2 copies of the CA cert bytes written to the hub on any change?
At least, that's how I read "old: a.details, new: details".

The fact that it is called a "Delta" makes it sound like it only includes what has actually change, but given things like a.details.Config, that looks more like it is the whole Old and whole New content.

Given later test context, it seems that details.Config isn't the complete config, but possibly only the updated values. it would still seem that we send 1 complete copy of the config on every change.

Reply:
We publish a _reference_ over the hub. Unless I misread the `SimpleHub` code this shouldn't be fanning out whole copies...

Reply:
func (h *StructuredHub) Publish(topic string, data interface{}) (<-chan struct{}, error) {
	if data == nil {
		data = make(map[string]interface{})
	}
	asMap, err := h.toStringMap(data)
	if err != nil {
		return nil, errors.Trace(err)
	}
...
func (h *StructuredHub) toStringMap(data interface{}) (map[string]interface{}, error) {
...
	bytes, err := h.marshaller.Marshal(data)


You are passing in a reference, but it is getting turned into a map[string]interface{} internally.

It would not be safe to pass references to memory around. Then we would have to worry about multithreaded access to everything.

@howbazaar would know best, but I'm pretty sure we send content over the hub, not references.

Reply:
That is `StructuredHub`. We're using `SimpleHub`.
```go
func (h *SimpleHub) Publish(topic string, data interface{}) <-chan struct{} {
	h.mutex.Lock()
	defer h.mutex.Unlock()

	done := make(chan struct{})
	wait := sync.WaitGroup{}

	for _, s := range h.subscribers {
		if s.topicMatcher(topic) {
			wait.Add(1)
			s.notify(
				&handlerCallback{
					topic: topic,
					data:  data,
					wg:    &wait,
				})
		}
	}

	go func() {
		wait.Wait()
		close(done)
	}()

	return done
}
```

Reply:
It is probably worth noting here that the hub is a local simple hub, that just is used within the cache object itself. The data that is passed into publish is an `interface{}` that is just passed through into the callbacks. There is no wire involved with this hub.